docs/proposals/04-new-core-gardener-cloud-apis.md

```diff
@@ -0,0 +1,698 @@
+# New `core.gardener.cloud/v1alpha1` APIs required to extract cloud-specific/OS-specific knowledge out of Gardener core
+
+## Table of Contents
+
+* [Table of Contents](#table-of-contents)
+* [Summary](#summary)
+* [Motivation](#motivation)
+    * [Goals](#goals)
+    * [Non-Goals](#non-goals)
+* [Proposal](#proposal)
+    * [`ProviderProfile` resource](#providerprofile-resource)
+    * [`Seed` resource](#seed-resource)
+    * [`SecretBinding` resource](#secretbinding-resource)
+    * [`Quota` resource](#quota-resource)
+    * [`BackupBucket` resource](#backupbucket-resource)
+    * [`BackupEntry` resource](#backupentry-resource)
+    * [`Shoot` resource](#shoot-resource)
+    * [`Plant` resource](#plant-resource)
+
+## Summary
+
+In [GEP-1](https://github.com/gardener/gardener/blob/master/docs/proposals/01-extensibility.md) we have proposed how to (re-)design Gardener to allow providers maintaining their provider-specific knowledge out of the core tree.
+Meanwhile, we have progressed a lot and are about to remove the [`CloudBotanist` interface](https://github.com/gardener/gardener/blob/master/pkg/operation/cloudbotanist/types.go) entirely.
+The only missing aspect that will allow providers to really maintain their code out of the core is to design new APIs.
+
+This proposal describes how the new `Shoot`, `Seed` etc. APIs will be re-designed to cope with the changes made with extensibility.
+We already have the new `core.gardener.cloud/v1alpha1` API group that will be the new default soon.
+
+## Motivation
+
+We want to allow providers to individually maintain their specific knowledge without the necessity to touch the Gardener core code.
+In order to achieve the same, we have to provide proper APIs.
+
+### Goals
+
+* Provide proper APIs to allow providers maintaining their code outside of the core codebase.
+* Do not complicate the APIs for end-users such that they can easily create, delete, and maintain shoot clusters.
+
+### Non-Goals
+
+* Let's try to not split everything up into too many different resources. Instead, let's try to keep all relevant information in the same resources when possible/appropriate.
+
+## Proposal
+
+In GEP-1 we already have proposed a first version for new `CloudProfile` and `Shoot` resources.
+In order to deprecate the existing/old `garden.sapcloud.io/v1beta1` API group (and remove it, eventually) we should move all existing resources to the new `core.gardener.cloud/v1alpha1` API group.
+
+### `ProviderProfile` resource
+
+Special note: The `CloudProfile` will be renamed to `ProviderProfile`.
+
+```yaml
+apiVersion: core.gardener.cloud/v1alpha1
+kind: ProviderProfile
```

Comment:
One problem with the current `CloudProfile` is that it's used for both mutation of fields (if not set) and policy (validation). `PodSecurityPolicy` suffers from the same problems. See this [issue](https://github.com/kubernetes/enhancements/issues/5#issuecomment-476332297) and [presentation](https://docs.google.com/presentation/d/1Kv6BSBNyLCyglMbK7e6tVOaDYe89LV2aHL2Hlb-9HX8/edit#slide=id.p).

I think that there might be a better solution for it.

Reply:
Do you have a concrete idea? Can you explain again what the problems are? I read the issue, but I didn't understand it quite well. I see that they consider moving PSP out of the core (we cannot do that), but I don't really understand the problems.

Reply:
Well, I think more or less the same comment as below applies here as well: Cannot happen in the first `v1alpha1` version. We need to allow extension controllers to plug into our API, and this was the main motivation for this proposal. Let's not do too many things together, this one thing is already huge and will cause lots of changes.